### ISSUE TITLE
Unwanted change of states

### ISSUE THREAD

**Issue by SMV818VMS** [pysb/pysb#244]
I am working with some polymeric structures and I am having some problems at the simulation time.

Let's consider a monomer R that is able to polymerize with other R's and a monomer T that can recognize and bind two R that are together.

```
from pysb import *
from pysb.integrate import odesolve
import pylab import linspace

Model()

Monomer('T', ['b1', 'b2'])              # Binding site to one R and to the other
Monomer('R', ['p1', 'p2', 'b'])        # p1 and p2 to polymerize and b to bind T
```

I create a 4-meric complex including the T bound, using a non-existent bound identifiers (20 and 21) to avoid conflicts and the rule of breaking the polymer between the two R binding T:

```
x = R(p1=None, p2=1, b=None) % R(p1=1, p2=2, b=20) % R(p1=2, p2=3, b=21) % R(p1=3, p2=None, b=None) % T(b1=20, b2=21)

Rule('break_polymer',
    R(p1=ANY, p2=1, b=2) % R(p1=1, p2=ANY, b=3) % T(b1=2, b2=3) >> R(p1=ANY, p2=None, b=None) + R(p1=None, p2=ANY, b=None) + T(b1=None, b2=None),
    Parameter('k', 2))
```

Finally, I create the observable representing number of bonds and the initial concentration of x:

```
Observable('bonds', R(p2=1) % R(p1=1))
Initial(x, Parameter('x_0', 3))
```

This model is accepted without any warning, but when I try to simulate it:

```
t = linspace(0, 60, 15)
ode_sim = odesolve(model, t)
```

I get the error:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-105-d01db2e6c951> in <module>()
      1 # ODE simulation
      2 t = linspace(0, 60, 15)
----> 3 ode_sim = odesolve(model, t)

/usr/local/lib/python2.7/dist-packages/pysb/integrate.pyc in odesolve(model, tspan, param_values, y0, integrator, cleanup, verbose, **integrator_options)
    226     sim = ScipyOdeSimulator(model, tspan=tspan, cleanup=cleanup,
    227                             verbose=verbose, **integrator_options)
--> 228     simres = sim.run(param_values=param_values, initials=y0)
    229     return simres.all
    230 

/usr/local/lib/python2.7/dist-packages/pysb/simulator/scipyode.pyc in run(self, tspan, param_values, initials)
    341         if initials is not None:
    342             self.initials = initials
--> 343         y0 = self.initials_list
    344         param_values = self.param_values
    345         if self.integrator == 'lsoda':

/usr/local/lib/python2.7/dist-packages/pysb/simulator/base.pyc in initials_list(self)
    158             _set_initials(self._initials.items())
    159         # Get remaining initials from the model itself
--> 160         _set_initials(self._model.initial_conditions)
    161 
    162         return y0

/usr/local/lib/python2.7/dist-packages/pysb/simulator/base.pyc in _set_initials(initials_source)
    139                 if si is None:
    140                     raise IndexError("Species not found in model: %s" %
--> 141                                      repr(cp))
    142                 # If this initial condition has already been set, skip it
    143                 if y0[si] != 0:

IndexError: Species not found in model: R(p1=None, p2=1, b=None) % R(p1=1, p2=2, b=20) % R(p1=2, p2=3, b=21) % R(p1=3, p2=None, b=None) % T(b1=20, b2=21)
```
This is because the interpreter seems to change the bond identifiers I assigned as `model.species` returns:

```
[R(p1=2, p2=3, b=1) % R(p1=3, p2=5, b=4) % R(p1=5, p2=None, b=None) % R(p1=None, p2=2, b=None) % T(b1=1, b2=4),
 R(p1=1, p2=None, b=None) % R(p1=None, p2=1, b=None),
 T(b1=None, b2=None)]
```

This complicates the application in general cases like trying to extend the usage to any polymeric size as the ending species will be re-structured without keeping the states I assign. 

Which could be a proper way to implement this?

Thanks in advance, 

Sam.

**Comment by jmuhlich** [comment]
Sorry, this is one of our most long-standing bugs -- see #13 . It's actually going to be quite a lot of work to fix it, but I appreciate your pain. We've done a bit of modeling with pores and chains and somehow managed to avoid this problem, possibly because we haven't tried to set an initial condition for a multimeric complex like your model does. Will you ultimately need to initialize the model with those complexes pre-formed, or is this just for testing?

**Comment by SMV818VMS** [comment]
Thanks for your interest in the problem. In my case the intuitive model requires the complex pre-formed and considering this,  I have been trying other ways to do the same.

The easiest way and less obfuscated option is to write a rule for binding T instead of starting bound and formulate the polymer formation in a different order. Instead of starting from the first element, start from the second and leave the first monomer to the end:

```
from pysb import *
from pysb.integrate import odesolve
import pylab import linspace

Model()

Monomer('T', ['b1', 'b2'])              # Binding site to one R and to the other
Monomer('R', ['p1', 'p2', 'b'])        # p1 and p2 to polymerize and b to bind T

# x = R(p1=None, p2=1, b=None) % R(p1=1, p2=2, b=None) % R(p1=2, p2=3, b=None) % R(p1=3, p2=None, b=None)
# order 1, 2, 3, 4
x = R(p1=1, p2=2, b=None) % R(p1=2, p2=3, b=None) % R(p1=3, p2=None, b=None) % R(p1=None, p2=1, b=None)
# Here is where it is different! order 2, 3, 4, 1

Rule('bind_polymer',
    R(p1=ANY, p2=1, b=None) % R(p1=1, p2=ANY, b=None) + T(b1=None, b2=None) >> R(p1=ANY, p2=1, b=2) % R(p1=1, p2=ANY, b=3) % T(b1=2, b2=3),
    Parameter('kbind', 1))

Rule('break_polymer',
    R(p1=ANY, p2=1, b=2) % R(p1=1, p2=ANY, b=3) % T(b1=2, b2=3) >> R(p1=ANY, p2=None, b=None) + R(p1=None, p2=ANY, b=None) + T(b1=None, b2=None),
    Parameter('kbreak', 1))

Observable('bonds', R(p2=1) % R(p1=1))

Initial(x, Parameter('x_0', 3))
Initial(T(b1=None, b2=None), Parameter('freeT', 1))
```

With this formulation the simulation can run:

```
# ODE simulation
>>> t = pl.linspace(0, 60, 10)
>>> ode_sim = odesolve(model, t)
>>> ode_sim['bonds']
array([ 9.        ,  6.0322176 ,  6.00004364,  6.00000015,  6.00000009,
        6.00000009,  6.00000009,  6.00000009,  6.00000009,  6.00000009]
```

I have to check in deep how the species are loaded but it seems that there is some kind of sorting based on the site states. 

This problem will occur using the polymer and pore functions provided by pysb.macros, but simply changing the last step of adding the last monomer and add after the addition of the first should work. It is not the most intuitive order to write it, but it works as expected.

Thanks again!

**Comment by jmuhlich** [comment]
The sorting is whatever BioNetGen outputs in its .net files -- Pysb just 
parses that output directly.

**Comment by alubbock** [comment]
Hi @SMV818VMS, I implemented a fix for the bond numbering issue #13 that @jmuhlich mentioned in his first reply. Are you still having any problems, or can I close this issue? Thanks.

**Comment by alubbock** [comment]
Closing due to lack of response. I believe this issue should now be fixed. Please re-open with further details if you're still experiencing issues.

Thanks.
